### PR TITLE
perf: support parallel assignment for V1 layouter

### PR DESCRIPTION
**Pull request by junyu0312**

### CODE REVIEW

File: halo2_proofs/src/circuit/floor_planner/single_pass.rs [hunk 1]
```diff
@@ -420,7 +67,8 @@ impl<'r, 'a, F: Field, CS: Assignment<F> + 'a> TableLayouter<F>
             return Err(Error::Synthesis); // TODO better error
         }
 
-        let entry = self.default_and_assigned.entry(column).or_default();
+        let mut default_and_assigned = self.default_and_assigned.lock().unwrap();
```
Comment (junyu0312) [inline halo2_proofs/src/circuit/floor_planner/single_pass.rs:70]: TODO

File: halo2_proofs/src/circuit/layouter.rs [hunk 2]
```diff
@@ -196,38 +198,44 @@ impl RegionShape {
...
         _to: &'v mut (dyn FnMut() -> Result<Assigned<F>, Error> + 'v),
     ) -> Result<Cell, Error> {
-        self.columns.insert(Column::<Any>::from(column).into());
-        self.row_count = cmp::max(self.row_count, offset + 1);
+        let mut region_shape = self.0.lock().unwrap();
+
+        region_shape
+            .columns
+            .insert(Column::<Any>::from(column).into());
+        region_shape.row_count = cmp::max(region_shape.row_count, offset + 1);
```
Comment (junyu0312) [inline halo2_proofs/src/circuit/layouter.rs:228]: TODO: perf

File: halo2_proofs/src/circuit/layouter.rs [hunk 3]
```diff
@@ -259,28 +271,32 @@ impl<F: Field> RegionLayouter<F> for RegionShape {
...
```
Comment (junyu0312) [inline halo2_proofs/src/circuit/layouter.rs:285]: TODO: perf

File: halo2_proofs/src/circuit/layouter.rs [hunk 4]
```diff
@@ -238,19 +246,23 @@ impl<F: Field> RegionLayouter<F> for RegionShape {
...
+        &self,
         _: &'v (dyn Fn() -> String + 'v),
         _: Column<Instance>,
         _: usize,
         advice: Column<Advice>,
         offset: usize,
     ) -> Result<(Cell, Option<F>), Error> {
-        self.columns.insert(Column::<Any>::from(advice).into());
-        self.row_count = cmp::max(self.row_count, offset + 1);
+        let mut region_shape = self.0.lock().unwrap();
```
Comment (junyu0312) [inline halo2_proofs/src/circuit/layouter.rs]: TODO: perf

File: halo2_proofs/src/dev.rs [hunk 5]
```diff
@@ -482,17 +491,39 @@ impl<F: Group + Field> Mul<F> for Value<F> {
 ///     } if current_k == 2,
 /// ));
 /// ```
-#[derive(Debug)]
+#[derive(Clone, Debug)]
```
Comment (junyu0312) [inline halo2_proofs/src/dev.rs]: remove TLS

File: halo2_proofs/src/circuit/floor_planner/v1.rs [hunk 6]
```diff
@@ -66,17 +72,20 @@ impl FloorPlanner for V1 {
         let mut plan = V1Plan::new(cs)?;
 
         // First pass: measure the regions within the circuit.
-        let mut measure = MeasurementPass::new();
+        let mut measure = MeasurementPassAssignment::new();
```
Comment (junyu0312) [inline halo2_proofs/src/circuit/floor_planner/v1.rs:75]: sort stable?

**Review by xgaozoyoe — APPROVED** [review]

**Pull request merged**